File: srcLegacy/Mapper_069.cpp
#include "Mapper_069.h"
#include <fstream>

// Track if we've logged bank changes (only log first few)
static int bankChangeCount = 0;

Mapper_069::Mapper_069(uint8_t prgBanks, uint8_t chrBanks)
    : Mapper(prgBanks, chrBanks) {
  vPRGRAM.resize(8192, 0);
  reset();
}

Mapper_069::~Mapper_069() {}

void Mapper_069::reset() {
  commandRegister = 0;

  // Initialize PRG banks
  prgBank[0] = 0;
  prgBank[1] = 0;
  prgBank[2] = 0;
  prgBank[3] = 0;
  prgRamEnable = false;
  prgRamSelect = false;

  // Initialize CHR banks to 1:1 mapping
  for (int i = 0; i < 8; i++) {
    chrBank[i] = i;
  }

  mirrorMode = MIRROR::VERTICAL;

  bIRQEnable = false;
  bIRQCounterEnable = false;
  bIRQActive = false;
  irqCounter = 0;
}

void Mapper_069::CountIRQ() {
  if (bIRQCounterEnable) {
    irqCounter--;
    if (irqCounter == 0xFFFF && bIRQEnable) {
      bIRQActive = true;
    }
  }
}

bool Mapper_069::cpuMapRead(uint16_t addr, uint32_t &mapped_addr) {
  uint32_t prgRomSize = nPRGBanks * 16384;

  // $6000-$7FFF: PRG RAM or ROM bank
  if (addr >= 0x6000 && addr <= 0x7FFF) {
    if (prgRamSelect) {
      // PRG RAM
      mapped_addr = 0xFFFFFFFF;
      return true;
    } else {
      // PRG ROM
      uint32_t bank = prgBank[0] & 0x3F;
      mapped_addr = ((bank * 8192) + (addr & 0x1FFF)) % prgRomSize;
      return true;
    }
  }

  // $8000-$9FFF: Switchable 8KB PRG ROM
  if (addr >= 0x8000 && addr <= 0x9FFF) {
    uint32_t bank = prgBank[1] & 0x3F;
    mapped_addr = ((bank * 8192) + (addr & 0x1FFF)) % prgRomSize;
    return true;
  }

  // $A000-$BFFF: Switchable 8KB PRG ROM
  if (addr >= 0xA000 && addr <= 0xBFFF) {
    uint32_t bank = prgBank[2] & 0x3F;
    mapped_addr = ((bank * 8192) + (addr & 0x1FFF)) % prgRomSize;
    return true;
  }

  // $C000-$DFFF: Switchable 8KB PRG ROM
  if (addr >= 0xC000 && addr <= 0xDFFF) {
    uint32_t bank = prgBank[3] & 0x3F;
    mapped_addr = ((bank * 8192) + (addr & 0x1FFF)) % prgRomSize;
    return true;
  }

  // $E000-$FFFF: Fixed to last 8KB bank
  if (addr >= 0xE000) {
    uint32_t lastBank = (nPRGBanks * 2) - 1; // Last 8KB bank
    mapped_addr = ((lastBank * 8192) + (addr & 0x1FFF)) % prgRomSize;
    return true;
  }

  return false;
}

bool Mapper_069::cpuMapWrite(uint16_t addr, uint32_t &mapped_addr) {
  return false;
}

bool Mapper_069::cpuMapWrite(uint16_t addr, uint32_t &mapped_addr,
                             uint8_t data) {
  // $6000-$7FFF: PRG RAM write
  if (addr >= 0x6000 && addr <= 0x7FFF) {
    if (prgRamSelect && prgRamEnable) {
      mapped_addr = 0xFFFFFFFF;
      return true;
    }
    return false;
  }

  // $8000-$9FFF: Command Register
  if (addr >= 0x8000 && addr <= 0x9FFF) {
    commandRegister = data & 0x0F;
    return false;
  }

  // $A000-$BFFF: Parameter Register
  if (addr >= 0xA000 && addr <= 0xBFFF) {
    switch (commandRegister) {
    case 0x0:
    case 0x1:
    case 0x2:
    case 0x3:
    case 0x4:
    case 0x5:
    case 0x6:
    case 0x7:
      // CHR Bank 0-7
      chrBank[commandRegister] = data;
      if (bankChangeCount < 20) {
        std::ofstream debugLog("nes_debug.log", std::ios::app);
        debugLog << "CHR Bank " << (int)commandRegister << " = " << (int)data
                 << std::endl;
        debugLog.close();
        bankChangeCount++;
      }
      break;

    case 0x8:
      // PRG Bank at $6000-$7FFF
      prgRamEnable = (data & 0x80) != 0;
      prgRamSelect = (data & 0x40) != 0;
      prgBank[0] = data & 0x3F;
      break;

    case 0x9:
      // PRG Bank at $8000-$9FFF
      prgBank[1] = data & 0x3F;
      break;

    case 0xA:
      // PRG Bank at $A000-$BFFF
      prgBank[2] = data & 0x3F;
      break;

    case 0xB:
      // PRG Bank at $C000-$DFFF
      prgBank[3] = data & 0x3F;
      break;

    case 0xC:
      // Mirroring
      switch (data & 0x03) {
      case 0:
        mirrorMode = MIRROR::VERTICAL;
        break;
      case 1:
        mirrorMode = MIRROR::HORIZONTAL;
        break;
      case 2:
        mirrorMode = MIRROR::ONESCREEN_LO;
        break;
      case 3:
        mirrorMode = MIRROR::ONESCREEN_HI;
        break;
      }
      break;

    case 0xD:
      // IRQ Control
      bIRQEnable = (data & 0x01) != 0;
      bIRQCounterEnable = (data & 0x80) != 0;
      if (!bIRQEnable) {
        bIRQActive = false;
      }
      break;

    case 0xE:
      // IRQ Counter Low
      irqCounter = (irqCounter & 0xFF00) | data;
      break;

    case 0xF:
      // IRQ Counter High
      irqCounter = (irqCounter & 0x00FF) | (data << 8);
      break;
    }
    return false;
  }

  return false;
}

bool Mapper_069::ppuMapRead(uint16_t addr, uint32_t &mapped_addr) {
  if (addr < 0x2000) {
    uint32_t chrRomSize = nCHRBanks * 8192;
    if (chrRomSize == 0)
      chrRomSize = 8192; // CHR RAM

    int bankIndex = (addr >> 10) & 0x07;
    uint32_t bank = chrBank[bankIndex];
    mapped_addr = ((bank * 1024) + (addr & 0x03FF)) % chrRomSize;
    return true;
  }
  return false;
}

bool Mapper_069::ppuMapWrite(uint16_t addr, uint32_t &mapped_addr) {
  // CHR RAM writes
  if (addr < 0x2000 && nCHRBanks == 0) {
    mapped_addr = addr;
    return true;
  }
  return false;
}

File: src/Mogiki.Core/Mappers/Mapper069.cs
using Mogiki.Core.Common;

namespace Mogiki.Core.Mappers;

/// <summary>
/// Mapper 069: Sunsoft FME-7 / Sunsoft 5B.
/// </summary>
public sealed class Mapper069 : Mapper
{
    private byte _commandRegister;
    private readonly byte[] _prgBank = new byte[4];
    private readonly byte[] _chrBank = new byte[8];

    private bool _prgRamEnable;
    private bool _prgRamSelect;
    private MirrorMode _mirrorMode = MirrorMode.Vertical;

    private bool _bIRQEnable;
    private bool _bIRQCounterEnable;
    private bool _bIRQActive;
    private ushort _irqCounter;

    public byte[] PrgRam { get; } = new byte[8192];

    public Mapper069(byte prgBanks, byte chrBanks)
        : base(prgBanks, chrBanks)
    {
        Reset();
    }

    public override MirrorMode Mirror => _mirrorMode;
    public override bool IrqState => _bIRQActive && _bIRQEnable;
    public override void IrqClear() => _bIRQActive = false;

    public override void Reset()
    {
        _commandRegister = 0;

        _prgBank[0] = 0;
        _prgBank[1] = 0;
        _prgBank[2] = 0;
        _prgBank[3] = 0;
        _prgRamEnable = false;
        _prgRamSelect = false;

        for (int i = 0; i < 8; i++)
        {
            _chrBank[i] = (byte)i;
        }

        _mirrorMode = MirrorMode.Vertical;
        _bIRQEnable = false;
        _bIRQCounterEnable = false;
        _bIRQActive = false;
        _irqCounter = 0;
    }

    public void CountIRQ()
    {
        if (_bIRQCounterEnable)
        {
            _irqCounter--;
            if (_irqCounter == 0xFFFF && _bIRQEnable)
            {
                _bIRQActive = true;
            }
        }
    }

    public override bool CpuMapRead(ushort addr, out uint mappedAddr)
    {
        uint prgRomSize = (uint)(nPRGBanks * 16384);

        // $6000-$7FFF: PRG RAM or ROM bank
        if (addr is >= 0x6000 and <= 0x7FFF)
        {
            if (_prgRamSelect)
            {
                mappedAddr = PrgRamSentinel;
                return true;
            }
            uint bank = (uint)(_prgBank[0] & 0x3F);
            mappedAddr = ((bank * 8192) + (uint)(addr & 0x1FFF)) % prgRomSize;
            return true;
        }

        // $8000-$9FFF
        if (addr is >= 0x8000 and <= 0x9FFF)
        {
            uint bank = (uint)(_prgBank[1] & 0x3F);
            mappedAddr = ((bank * 8192) + (uint)(addr & 0x1FFF)) % prgRomSize;
            return true;
        }

        // $A000-$BFFF
        if (addr is >= 0xA000 and <= 0xBFFF)
        {
            uint bank = (uint)(_prgBank[2] & 0x3F);
            mappedAddr = ((bank * 8192) + (uint)(addr & 0x1FFF)) % prgRomSize;
            return true;
        }

        // $C000-$DFFF
        if (addr is >= 0xC000 and <= 0xDFFF)
        {
            uint bank = (uint)(_prgBank[3] & 0x3F);
            mappedAddr = ((bank * 8192) + (uint)(addr & 0x1FFF)) % prgRomSize;
            return true;
        }

        // $E000-$FFFF
        if (addr >= 0xE000)
        {
            uint lastBank = (uint)((nPRGBanks * 2) - 1);
            mappedAddr = ((lastBank * 8192) + (uint)(addr & 0x1FFF)) % prgRomSize;
            return true;
        }

        mappedAddr = 0;
        return false;
    }

    public override bool CpuMapWrite(ushort addr, out uint mappedAddr, byte data)
    {
        if (addr is >= 0x6000 and <= 0x7FFF)
        {
            if (_prgRamSelect && _prgRamEnable)
            {
                PrgRam[addr & 0x1FFF] = data;
                mappedAddr = PrgRamSentinel;
                return true;
            }
            mappedAddr = 0;
            return false;
        }

        // $8000-$9FFF: Command Register
        if (addr is >= 0x8000 and <= 0x9FFF)
        {
            _commandRegister = (byte)(data & 0x0F);
            mappedAddr = 0;
            return false;
        }

        // $A000-$BFFF: Parameter Register
        if (addr is >= 0xA000 and <= 0xBFFF)
        {
            switch (_commandRegister)
            {
                case <= 0x7:
                    _chrBank[_commandRegister] = data;
                    break;

                case 0x8:
                    _prgRamEnable = (data & 0x80) != 0;
                    _prgRamSelect = (data & 0x40) != 0;
                    _prgBank[0] = (byte)(data & 0x3F);
                    break;

                case 0x9:
                    _prgBank[1] = (byte)(data & 0x3F);
                    break;

                case 0xA:
                    _prgBank[2] = (byte)(data & 0x3F);
                    break;

                case 0xB:
                    _prgBank[3] = (byte)(data & 0x3F);
                    break;

                case 0xC:
                    _mirrorMode = (data & 0x03) switch
                    {
                        0 => MirrorMode.Vertical,
                        1 => MirrorMode.Horizontal,
                        2 => MirrorMode.OneScreenLo,
                        _ => MirrorMode.OneScreenHi
                    };
                    break;

                case 0xD:
                    _bIRQEnable = (data & 0x01) != 0;
                    _bIRQCounterEnable = (data & 0x80) != 0;
                    if (!_bIRQEnable)
                    {
                        _bIRQActive = false;
                    }
                    break;

                case 0xE:
                    _irqCounter = (ushort)((_irqCounter & 0xFF00) | data);
                    break;

                case 0xF:
                    _irqCounter = (ushort)((_irqCounter & 0x00FF) | (data << 8));
                    break;
            }
            mappedAddr = 0;
            return false;
        }

        mappedAddr = 0;
        return false;
    }

    public override bool CpuMapWrite(ushort addr, out uint mappedAddr)
    {
        mappedAddr = 0;
        return false;
    }

    public override bool PpuMapRead(ushort addr, out uint mappedAddr)
    {
        if (addr < 0x2000)
        {
            uint chrRomSize = (uint)(nCHRBanks * 8192);
            if (chrRomSize == 0) chrRomSize = 8192;

            int bankIndex = (addr >> 10) & 0x07;
            uint bank = _chrBank[bankIndex];
            mappedAddr = ((bank * 1024) + (uint)(addr & 0x03FF)) % chrRomSize;
            return true;
        }
        mappedAddr = 0;
        return false;
    }

    public override bool PpuMapWrite(ushort addr, out uint mappedAddr)
    {
        if (addr < 0x2000 && nCHRBanks == 0)
        {
            mappedAddr = addr;
            return true;
        }
        mappedAddr = 0;
        return false;
    }
}
